Tests/NIOTests/LengthFieldBasedFrameDecoderTest.swift

```diff
@@ -0,0 +1,91 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+//
+import XCTest
+
+struct fixedLengthFrameDecoderHelper<T: FixedWidthInteger> {
+    static fileprivate func fixedLengthFrameHelper(_ channel: EmbeddedChannel) throws {
+        XCTAssertNoThrow(_ = try channel.pipeline.add(handler: LengthFieldBasedFrameDecoder<T>()).wait())
+        
+        let testStrings = ["try", "swift", "san", "jose"]
+        var buffer = channel.allocator.buffer(capacity:1024)
+        for testString in testStrings {
+            buffer.write(integer: T(testString.count))
+            buffer.write(string: testString)
+        }
+        
+        for testString in testStrings {
+            XCTAssertTrue(try channel.writeInbound(buffer))
+            
+            guard var result: ByteBuffer = channel.readInbound() else {
+                XCTFail("Failed to read buffer for \(T.self)")
+                return
+            }
+            let testResult = result.readString(length: (result.readableBytes))
+            XCTAssertEqual(testResult, testString)
+        }
+        XCTAssertFalse(try channel.finish())
+    }
+}
+
+public class LengthFieldBasedFrameDecoderTest: XCTestCase {
```

Comment:
we need tests for the edge cases, for example:

- negative lengths
- absurdly large lengths
- length and then before receiving the appropriate number of bytes we get an EOF
- EOF whilst waiting for length

Reply:
@weissi I'm having trouble with the last 2 tests. Could you take a look at `testEOFBeforeReceivingAppropriateNumberOfBytes` & `testEOFWhileWaitingForLength` and provide some pointers? Thanks!

Reply:
hi @blossomwoo sure, that's not obvious at all actually. Let me run you through what I think we need to do to get this in asap. We're really close!

If anything doesn't make sense, please just at-mention me and ask for help. And sorry this is quite a bit of stuff actually.

So first we should make some very minor changes to the actual code:

1) we should create a new `private func handleLeftOverData()` which contains the logic to fire the error if there's left-over stuff
2) we should call `self.handleLeftOverData()` from both `func handlerRemoved()` and `func channelInactive()`

The reason we should do it in both `channelInactive` (called when the channel gets inactive, in TCP that means the socket is closed) and `handlerRemoved` (called when the handler gets removed from the pipeline) is: In probably most cases, you'll see `channelInactive` followed by `handlerRemoved` because we remove all handlers because we tear the pipeline down. However, a handler might also be removed in normal operation in which case we would also like to fire the error. The easiest way of achieving this is to call the left-over logic in both. Note: we need to set the left-over bytes to nothing after firing the error.

regarding the tests, what probably works best is to create a new channel handler called something like

    private final class LeftOverCheckHandler: ChannelInboundHandler {
        typealias InboundIn = ByteBuffer
        typealias InboundOut = ByteBuffer

        enum LeftOverCheckHandlerError: Error {
            case wrongErrorReceived(Error)
        }

        private var haveSeenLeftOvers = false
        private let leftOversReceivedPromise: EventLoopPromise<ByteBuffer>

        init(leftOversReceivedPromise: EventLoopPromise<ByteBuffer >) {
            self.leftOversReceivedPromise = leftOversReceivedPromise
        }

        func errorCaught(ctx: ChannelHandlerContext, error: Error) {
            XCTAssertFalse(self.haveSeenLeftOvers)
            self.haveSeenLeftOvers = true

            switch error {
                case let error as NIOLengthFieldBasedFrameDecoderError:
                    switch error {
                        case .bytesLeftOver(let bytes):
                            self.leftOversReceivedPromise.succeed(value: bytes)
                        case .invalidFrameLength:
                            self.leftOversReceivedPromise.fail(error: LeftOverCheckHandlerError.wrongErrorReceived(error))
                    }
                  default:
                    self.leftOversReceivedPromise.fail(error: LeftOverCheckHandlerError.wrongErrorReceived(error))
                }
         }
     }

and you can just add this channel handler to the channel (using the following code)

```
let leftOversReceivedPromise: EventLoopPromise<ByteBuffer> = embedded.eventLoop.newPromise()
XCTAssertNoThrow(try embedded.pipeline.add(handler: LeftOverCheckHandler(leftOversReceivedPromise: leftOversReceivedPromise)).wait())
```

and now you can do a bunch of test cases:

1) just fire `channelInactive` through the pipeline manually: `embedded.pipeline.fireChannelInactive()` then verify that the left over bytes got received correctly by the `LeftOverCheckHandler`:

    XCTAssertEqual(whateverLeftOverBytesYouExpectAsByteBuffer,
                                 leftOversReceivedPromise.futureResult.wait()!)

2) call `embedded.close().wait()` and then spin the event loop (`embedded.eventloop.run()`) and then verify the same thing as in (1)
3) left the embedded channel up and running but remove the handler (`embedded.pipeline.remove(handler: theHandlerInstanceOfYourFixedLengthDecoderHandler)`) and then verify the same thing as in (1) and (2)

(disclaimer: I tested/compiled none of the code in here so there might be some minor mistakes, bugs, typos, etc)